### ISSUE TITLE
0.0.2: Canonical FastAPI knowledge write path and route surface

### ISSUE THREAD

**Issue by dtubb** [dtubb/fichero#364]
Implement canonical FastAPI mutation paths for sources, entities, claims, and claim-links. Ensure no direct client datastore writes for knowledge operations.\n\nAcceptance:\n- Dedicated knowledge routes are registered in API tier routing\n- Service-layer mutation methods enforce referential integrity\n- Contract tests pass for SwiftUI + MCP clients

**Comment by dtubb** [comment]
Roadmap dependency note: cannot start before none. Canonical contracts/gates and full matrix are in docs/0.0.2-planning/AUTHORITATIVE_ROADMAP_SPEC.md (Section 5).

**Comment by dtubb** [comment]
Backend update (0.0.2 branch): investigated and fixed sources route + security blockers.

What was fixed:
-  route behavior corrected (Document field mapping + tests passing)
- ✓ OpenAPI schema exported to /Users/danieltubb/code/fichero-0.0.2/fichero-api/tests/contracts/openapi.json
✓ Endpoints list exported to /Users/danieltubb/code/fichero-0.0.2/fichero-api/tests/contracts/endpoints.json

Exported 267 endpoints across 23 resources:
  activity: 12 endpoints
  artifacts: 5 endpoints
  batches: 10 endpoints
  chat: 9 endpoints
  documents: 13 endpoints
  folders: 5 endpoints
  graph: 5 endpoints
  health: 1 endpoints
  hermeneutics: 20 endpoints
  ingest: 3 endpoints
  knowledge-graph: 30 endpoints
  mcp: 8 endpoints
  mind-palace: 31 endpoints
  models: 3 endpoints
  providers: 19 endpoints
  research: 28 endpoints
  search: 14 endpoints
  settings: 3 endpoints
  sources: 5 endpoints
  stats: 1 endpoints
  storage: 11 endpoints
  workflow-execution: 16 endpoints
  workflows: 15 endpoints
- Additional file comments: <none>Writing data to file Client.swift...
Writing data to file Types.swift...
Build complete! (59.32s)
✅ Synced backend OpenAPI schema into Swift client package (fichero-api-client)
Validating Python/Swift model sync...
  Python: src/fichero/workflows/types.py
  Swift files:
    - fichero-swiftui/fichero-swiftui/Models/WorkflowTypes.swift
    - fichero-swiftui/fichero-swiftui/Models/WorkflowSupportTypes.swift
    - fichero-swiftui/fichero-swiftui/Models/WorkflowToolTypes.swift

✅ All models are in sync! updated to prefer project-local  before external  to avoid cross-worktree imports
- Added library-path validation guardrails in API ()
- Added explicit redirect-safe HTTP fetch flow for research tools (blocks SSRF redirect bypass)
- Hardened URL safety checks for traversal/query redirect gadget patterns
- Added PyKEEN compatibility shim for  on installed version
- Added ML-training sensitivity filter in PyKEEN triple builder (claims default to confidential unless explicitly public; API-created claims are marked public)

Verification:
- ============================= test session starts ==============================
platform darwin -- Python 3.12.13, pytest-9.0.2, pluggy-1.6.0
rootdir: /Users/danieltubb/code/fichero-0.0.2/fichero-api
configfile: pyproject.toml
plugins: langsmith-0.7.22, anyio-4.13.0
collected 1249 items

fichero-api/tests/unit/test_action_store.py ........                     [  0%]
fichero-api/tests/unit/test_activity.py ..................               [  2%]
fichero-api/tests/unit/test_activity_enhancements.py ............        [  3%]
fichero-api/tests/unit/test_api.py ..................................... [  6%]
................                                                         [  7%]
fichero-api/tests/unit/test_api_providers.py ......................      [  9%]
fichero-api/tests/unit/test_backend_integration.py ....................  [ 10%]
fichero-api/tests/unit/test_bookmarks.py ...sss...s.....sss..s...        [ 12%]
fichero-api/tests/unit/test_contract_models.py ...........               [ 13%]
fichero-api/tests/unit/test_data_layer_deps.py ......................    [ 15%]
fichero-api/tests/unit/test_db.py ...................................... [ 18%]
........                                                                 [ 18%]
fichero-api/tests/unit/test_feature_tier_routing.py ...                  [ 19%]
fichero-api/tests/unit/test_graph_exploration.py ...............         [ 20%]
fichero-api/tests/unit/test_graph_reasoning.py ......................    [ 22%]
fichero-api/tests/unit/test_hermeneutics_api.py ...........              [ 22%]
fichero-api/tests/unit/test_hermeneutics_security.py .....               [ 23%]
fichero-api/tests/unit/test_ingest_module.py ........................... [ 25%]
................................s...................                     [ 29%]
fichero-api/tests/unit/test_integration_security.py ............s        [ 30%]
fichero-api/tests/unit/test_integrations.py .......................      [ 32%]
fichero-api/tests/unit/test_knowledge_graph_api.py ..................... [ 34%]
..                                                                       [ 34%]
fichero-api/tests/unit/test_knowledge_graph_security.py s.sss...         [ 35%]
fichero-api/tests/unit/test_knowledge_models.py ........................ [ 36%]
............                                                             [ 37%]
fichero-api/tests/unit/test_llm.py ..............sssss..ss               [ 39%]
fichero-api/tests/unit/test_mcp_manager.py ............................. [ 42%]
....                                                                     [ 42%]
fichero-api/tests/unit/test_mcp_server.py .............................. [ 44%]
...                                                                      [ 45%]
fichero-api/tests/unit/test_mcp_tools.py ................                [ 46%]
fichero-api/tests/unit/test_migrations.py ...............                [ 47%]
fichero-api/tests/unit/test_mind_palace_api.py .............             [ 48%]
fichero-api/tests/unit/test_mind_palace_security.py ........             [ 49%]
fichero-api/tests/unit/test_model_comparison.py ...................      [ 50%]
fichero-api/tests/unit/test_multilingual.py ............................ [ 53%]
.................                                                        [ 54%]
fichero-api/tests/unit/test_new_data_layer.py ..................         [ 55%]
fichero-api/tests/unit/test_node_cache.py ........................       [ 57%]
fichero-api/tests/unit/test_orchestration_policy.py .................... [ 59%]
...........                                                              [ 60%]
fichero-api/tests/unit/test_providers.py ............................... [ 62%]
.........                                                                [ 63%]
fichero-api/tests/unit/test_pykeen.py .........................          [ 65%]
fichero-api/tests/unit/test_research_agents_api.py ..................... [ 67%]
.                                                                        [ 67%]
fichero-api/tests/unit/test_research_ssrf_security.py .................. [ 68%]
...................................                                      [ 71%]
fichero-api/tests/unit/test_review_queue.py ...............              [ 72%]
fichero-api/tests/unit/test_search_explain.py ......................     [ 74%]
fichero-api/tests/unit/test_sources.py .......                           [ 74%]
fichero-api/tests/unit/test_storage.py ............................      [ 77%]
fichero-api/tests/unit/test_tasks.py ...................                 [ 78%]
fichero-api/tests/unit/test_tool_registry.py ...................         [ 80%]
fichero-api/tests/unit/test_workflow_api_direct.py ...                   [ 80%]
fichero-api/tests/unit/test_workflow_api_verification.py ...             [ 80%]
fichero-api/tests/unit/test_workflow_crud_operations.py ....             [ 81%]
fichero-api/tests/unit/test_workflow_executor.py ....................... [ 82%]
........                                                                 [ 83%]
fichero-api/tests/unit/test_workflow_import_export.py ....               [ 83%]
fichero-api/tests/unit/test_workflow_rename_duplicate.py ....            [ 84%]
fichero-api/tests/unit/test_workflow_tools.py .......................... [ 86%]
..........                                                               [ 87%]
fichero-api/tests/unit/test_workflows.py ............................    [ 89%]
fichero-api/tests/unit/workflows/test_agent_tool.py ........             [ 89%]
fichero-api/tests/unit/workflows/test_chaining.py ...................... [ 91%]
...............                                                          [ 92%]
fichero-api/tests/unit/workflows/test_checkpointer.py .................. [ 94%]
                                                                         [ 94%]
fichero-api/tests/unit/workflows/test_checkpointer_integration.py ...... [ 94%]
                                                                         [ 94%]
fichero-api/tests/unit/workflows/test_mcp_tools.py ..........            [ 95%]
fichero-api/tests/unit/workflows/test_multi_agent.py ..............      [ 96%]
fichero-api/tests/unit/workflows/test_workflow_store.py ................ [ 97%]
.........................                                                [100%]

=============================== warnings summary ===============================
fichero-api/tests/unit/test_research_ssrf_security.py:282
  /Users/danieltubb/code/fichero-0.0.2/fichero-api/tests/unit/test_research_ssrf_security.py:282: PytestUnknownMarkWarning: Unknown pytest.mark.slow - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.slow

tests/unit/test_ingest_module.py::TestTextExtraction::test_text_extraction_from_pdf_file
tests/unit/test_ingest_module.py::TestTextExtraction::test_text_extraction_from_pdf_file
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute

tests/unit/test_ingest_module.py::TestTextExtraction::test_text_extraction_from_pdf_file
tests/unit/test_ingest_module.py::TestTextExtraction::test_text_extraction_from_pdf_file
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyObject has no __module__ attribute

tests/unit/test_ingest_module.py::TestTextExtraction::test_text_extraction_from_pdf_file
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type swigvarlink has no __module__ attribute

tests/unit/test_knowledge_graph_api.py::test_embed_claims_and_semantic_search
tests/unit/test_knowledge_graph_api.py::test_embed_entities_and_semantic_search
tests/unit/test_knowledge_graph_api.py::test_find_similar_claims
  /Users/danieltubb/code/fichero-0.0.2/fichero-api/src/fichero/db.py:811: UserWarning: The model intfloat/multilingual-e5-large now uses mean pooling instead of CLS embedding. In order to preserve the previous behaviour, consider either pinning fastembed version to 0.5.1 or using `add_custom_model` functionality.
    self._embedder = TextEmbedding(

tests/unit/test_knowledge_graph_api.py::test_generate_pykeen_predictions_persists_run_with_metadata
tests/unit/test_knowledge_graph_api.py::test_generate_pykeen_predictions_persists_run_with_metadata
  /Users/danieltubb/code/fichero/fichero-api/.briefcase-venv/lib/python3.12/site-packages/pykeen/training/slcwa.py:69: DeprecationWarning: Training instances are always shuffled.
    dataset=cls.from_triples_factory(

tests/unit/test_knowledge_graph_api.py::test_generate_pykeen_predictions_persists_run_with_metadata
  /Users/danieltubb/code/fichero/fichero-api/.briefcase-venv/lib/python3.12/site-packages/torch/utils/data/dataloader.py:775: UserWarning: 'pin_memory' argument is set as true but not supported on MPS now, device pinned memory won't be used.
    super().__init__(loader)

tests/unit/workflows/test_chaining.py::TestEvaluateCondition::test_simple_equality
  /Users/danieltubb/code/fichero-0.0.2/fichero-api/src/fichero/workflows/chaining.py:392: DeprecationWarning: ast.Num is deprecated and will be removed in Python 3.14; use ast.Constant instead
    ast_num = getattr(ast, "Num", None)

tests/unit/workflows/test_chaining.py::TestEvaluateCondition::test_simple_equality
  /Users/danieltubb/code/fichero-0.0.2/fichero-api/src/fichero/workflows/chaining.py:393: DeprecationWarning: ast.Str is deprecated and will be removed in Python 3.14; use ast.Constant instead
    ast_str = getattr(ast, "Str", None)

tests/unit/workflows/test_chaining.py::TestEvaluateCondition::test_simple_equality
  /Users/danieltubb/code/fichero-0.0.2/fichero-api/src/fichero/workflows/chaining.py:394: DeprecationWarning: ast.NameConstant is deprecated and will be removed in Python 3.14; use ast.Constant instead
    ast_name_constant = getattr(ast, "NameConstant", None)

tests/unit/workflows/test_chaining.py::TestEvaluateCondition::test_simple_equality
tests/unit/workflows/test_chaining.py::TestEvaluateCondition::test_inequality
tests/unit/workflows/test_chaining.py::TestEvaluateCondition::test_numeric_comparison
tests/unit/workflows/test_chaining.py::TestEvaluateCondition::test_false_condition
tests/unit/workflows/test_chaining.py::TestEvaluateCondition::test_missing_path_condition
  /Users/danieltubb/code/fichero-0.0.2/fichero-api/src/fichero/workflows/chaining.py:405: DeprecationWarning: ast.Num is deprecated and will be removed in Python 3.14; use ast.Constant instead
    elif ast_num is not None and isinstance(

tests/unit/workflows/test_chaining.py::TestEvaluateCondition::test_simple_equality
tests/unit/workflows/test_chaining.py::TestEvaluateCondition::test_inequality
tests/unit/workflows/test_chaining.py::TestEvaluateCondition::test_numeric_comparison
tests/unit/workflows/test_chaining.py::TestEvaluateCondition::test_false_condition
tests/unit/workflows/test_chaining.py::TestEvaluateCondition::test_missing_path_condition
  /Users/danieltubb/code/fichero-0.0.2/fichero-api/src/fichero/workflows/chaining.py:410: DeprecationWarning: ast.Str is deprecated and will be removed in Python 3.14; use ast.Constant instead
    elif ast_str is not None and isinstance(

tests/unit/workflows/test_chaining.py::TestEvaluateCondition::test_simple_equality
tests/unit/workflows/test_chaining.py::TestEvaluateCondition::test_inequality
tests/unit/workflows/test_chaining.py::TestEvaluateCondition::test_numeric_comparison
tests/unit/workflows/test_chaining.py::TestEvaluateCondition::test_false_condition
tests/unit/workflows/test_chaining.py::TestEvaluateCondition::test_missing_path_condition
  /Users/danieltubb/code/fichero-0.0.2/fichero-api/src/fichero/workflows/chaining.py:415: DeprecationWarning: ast.NameConstant is deprecated and will be removed in Python 3.14; use ast.Constant instead
    elif ast_name_constant is not None and isinstance(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========== 1228 passed, 21 skipped, 30 warnings in 68.57s (0:01:08) =========== → **1228 passed, 21 skipped**
- Smoke test with backend started via  confirms  GET/POST works with valid library header.

Next:
- Stage/commit these changes and open PR against main once Daniel approves.

**Comment by dtubb** [comment]
Implemented via #454 and merged into 0.0.2. Sources route surface, security hardening, and unit test coverage are now in place on 0.0.2.
